sdk/helper/keysutil/policy.go

```diff
@@ -2392,3 +2395,194 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
 	return base64.StdEncoding.EncodeToString(wrappedKeys), nil
 }
+
+func (p *Policy) CreateCSR(keyVersion int, csrTemplate *x509.CertificateRequest) ([]byte, error) {
+	if !p.Type.SigningSupported() {
+		return nil, errutil.UserError{Err: fmt.Sprintf("key type '%s' does not support signing", p.Type)}
+	}
+
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	if keyEntry.IsPrivateKeyMissing() {
+		return nil, errutil.UserError{Err: "private key not imported for key version selected"}
+	}
+
+	csrTemplate.Signature = nil
+	csrTemplate.SignatureAlgorithm = x509.UnknownSignatureAlgorithm
+
+	key, err := p.getPrivateKey(&keyEntry)
+	if err != nil {
+		return nil, err
+	}
+
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, csrTemplate, key)
+	if err != nil {
+		return nil, fmt.Errorf("could not create the CSR: %w", err)
+	}
+
+	pemCSR := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: csrBytes,
+	})
+
+	return pemCSR, nil
+}
+
+func (p *Policy) getPrivateKey(keyEntry *KeyEntry) (crypto.Signer, error) {
+	switch p.Type {
+	case KeyType_ECDSA_P256, KeyType_ECDSA_P384, KeyType_ECDSA_P521:
+		curve, err := p.getECDSAKeyCurve()
+		if err != nil {
+			return nil, err
+		}
+		key := &ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{
+				Curve: curve,
+				X:     keyEntry.EC_X,
+				Y:     keyEntry.EC_Y,
+			},
+			D: keyEntry.EC_D,
+		}
+		return key, nil
+	case KeyType_ED25519:
+		if p.Derived {
+			return nil, errutil.UserError{Err: "operation not supported on keys with derivation enabled"}
+		}
+		return ed25519.PrivateKey(keyEntry.Key), nil
+	case KeyType_RSA2048, KeyType_RSA3072, KeyType_RSA4096:
+		return keyEntry.RSAKey, nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("selected key type '%s' does not support signing", p.Type.String())}
+	}
+}
+
+func (p *Policy) getECDSAKeyCurve() (elliptic.Curve, error) {
+	switch p.Type {
+	case KeyType_ECDSA_P256:
+		return elliptic.P256(), nil
+	case KeyType_ECDSA_P384:
+		return elliptic.P384(), nil
+	case KeyType_ECDSA_P521:
+		return elliptic.P521(), nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("key type '%s' does not have a curve parameter", p.Type)}
+	}
+}
+
+func (p *Policy) PersistCertificateChain(ctx context.Context, storage logical.Storage, keyVersion int, certificateChain []*x509.Certificate) error {
+	// validate that the certificate chain has at least one certifiate, the leaf certificate
+	if len(certificateChain) == 0 {
+		return errutil.UserError{Err: "expected at least one certificate in the certificate chain"}
+	}
+
+	// validate that the first element in the certificate chain is a leaf certificate
+	if !certificateChain[0].BasicConstraintsValid || certificateChain[0].IsCA {
+		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
+	}
+
+	// validate that the last element in the certificate chain is a CA certificate
```

Comment:
Ah, I do think we should collapse these now.

```go
if len(certificateChain) > 1 {
	// validate that the certificate chain contains only one leaf certificate
	for _, certificate := range certificateChain[1 : len(certificateChain)] {
		if !certificate.BasicConstraintsValid || !certificate.IsCA {
			return errutil.UserError{Err: "could not validate if there is only one leaf certificate in the provided chain"}
		}
	}
}
```

This will fix the panic:

```
2024-09-25T11:26:06.715-0400 [INFO]  TestTransit_Export_CertificateChain: http2: panic serving 127.0.0.1:47784: runtime error: slice bounds out of range [1:0]
goroutine 663 [running]:
net/http.(*http2serverConn).runHandler.func1()
	/usr/local/go/src/net/http/h2_bundle.go:6185 +0x145
panic({0x2bfc8c0?, 0xc001624450?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
github.com/openbao/openbao/sdk/v2/helper/keysutil.(*Policy).PersistCertificateChain(0xc001615900?, {0x362fbc0?, 0xc00161bec0?}, {0x3632620?, 0xc000bef630?}, 0x0?, {0xc000a922d0?, 0x0?, 0x0?})
	/home/ascheel/GitHub/cipherboy/openbao/sdk/helper/keysutil/policy.go:2492 +0x5ba
github.com/openbao/openbao/builtin/logical/transit.(*backend).pathImportCertChainWrite(0xc00032f030, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860, 0xc001663190)
	/home/ascheel/GitHub/cipherboy/openbao/builtin/logical/transit/path_certificates.go:200 +0x5a6
github.com/openbao/openbao/sdk/v2/framework.(*Backend).HandleRequest(0xc000cec0e0, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/sdk/framework/backend.go:293 +0xac8
github.com/openbao/openbao/vault.(*Router).routeCommon(0xc00088c140, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860, 0x0)
	/home/ascheel/GitHub/cipherboy/openbao/vault/router.go:775 +0x153c
github.com/openbao/openbao/vault.(*Router).Route(...)
	/home/ascheel/GitHub/cipherboy/openbao/vault/router.go:552
github.com/openbao/openbao/vault.(*Core).doRouting(0xc00121b320?, {0x362fbc0?, 0xc00161bec0?}, 0xc000200690?)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:823 +0x26
github.com/openbao/openbao/vault.(*Core).handleRequest(0xc00073a708, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:964 +0xf73
github.com/openbao/openbao/vault.(*Core).handleCancelableRequest(0xc00073a708, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:697 +0x11ff
github.com/openbao/openbao/vault.(*Core).switchedLockHandleRequest(0xc00073a708, {0x362fbc0, 0xc00161b620}, 0xc0000fd860, 0x1)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:529 +0x605
github.com/openbao/openbao/vault.(*Core).HandleRequest(...)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:482
github.com/openbao/openbao/http.request(0xc00073a708, {0x3620df0, 0xc00161b530}, 0xc0006bd0e0, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/http/handler.go:819 +0x7f
github.com/openbao/openbao/http.handler.handleLogical.handleLogicalInternal.func57({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/logical.go:356 +0xa6
net/http.HandlerFunc.ServeHTTP(0xc00073a708?, {0x3620df0?, 0xc00161b530?}, 0xc00161bda0?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.handleRequestForwarding.func35({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/handler.go:753 +0x143
net/http.HandlerFunc.ServeHTTP(0xc000634c30?, {0x3620df0?, 0xc00161b530?}, 0x6d7e88?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
net/http.(*ServeMux).ServeHTTP(0xc0008fe0c0?, {0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/usr/local/go/src/net/http/server.go:2688 +0x1ad
github.com/openbao/openbao/http.handler.wrapHelpHandler.func47({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/help.go:28 +0xfd
net/http.HandlerFunc.ServeHTTP(0x0?, {0x3620df0?, 0xc00161b530?}, 0xc001115428?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.wrapCORSHandler.func48({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/cors.go:32 +0x389
net/http.HandlerFunc.ServeHTTP(0xc00073a708?, {0x3620df0?, 0xc00161b530?}, 0xc000c3cd20?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.rateLimitQuotaWrapping.func49({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/util.go:156 +0x9bf
net/http.HandlerFunc.ServeHTTP(0xc00161b440?, {0x3620df0?, 0xc00161b530?}, 0xc000ce2e70?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.wrapGenericHandler.func1({0x3620c70, 0xc000a92238}, 0xc000c15e60)
	/home/ascheel/GitHub/cipherboy/openbao/http/handler.go:421 +0xd35
net/http.HandlerFunc.ServeHTTP(0xc000c15d40?, {0x3620c70?, 0xc000a92238?}, 0xc00161b470?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.wrapMaxRequestSizeHandler.func50({0x3620c70, 0xc000a92238}, 0xc000c15d40)
	/home/ascheel/GitHub/cipherboy/openbao/http/util.go:58 +0x122
net/http.HandlerFunc.ServeHTTP(0xc001658000?, {0x3620c70?, 0xc000a92238?}, 0x0?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.PrintablePathCheckHandler.func51({0x3620c70, 0xc000a92238}, 0xc000c15d40)
	/home/ascheel/go/pkg/mod/github.com/hashicorp/go-cleanhttp@v0.5.2/handlers.go:42 +0x8f
net/http.HandlerFunc.ServeHTTP(0xd?, {0x3620c70?, 0xc000a92238?}, 0x524468?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
net/http.serverHandler.ServeHTTP({0x79c900?}, {0x3620c70?, 0xc000a92238?}, 0xc000cec000?)
	/usr/local/go/src/net/http/server.go:3142 +0x8e
net/http.initALPNRequest.ServeHTTP({{0x362fbc0?, 0xc001456690?}, 0xc000f16388?, {0xc0005361e0?}}, {0x3620c70, 0xc000a92238}, 0xc000c15d40)
	/usr/local/go/src/net/http/server.go:3750 +0x231
net/http.(*http2serverConn).runHandler(0x4e47480?, 0x0?, 0x0?, 0xc0009b4000?)
	/usr/local/go/src/net/http/h2_bundle.go:6192 +0xbb
created by net/http.(*http2serverConn).scheduleHandler in goroutine 497
	/usr/local/go/src/net/http/h2_bundle.go:6127 +0x21d
2024-09-25T11:26:06.715-0400 [INFO]  TestTransit_Export_CertificateChain: cleaning up vault cluster
```

Reply:
Nice catch. Forgot about this.